### ISSUE TITLE
Can coap_server recognize network disconnected? 

### ISSUE THREAD

**Issue by hongnguyen-tma** [obgm/libcoap#172]
Hi
From previous issues, I 've learnt that a coap_client can recognized network disconnected when it exceeds message max-retransmission. Is there a way for a coap_server to recognize that network has been disconnected too?  Please share with me if you know how to test it.

(I have tried by turning off client right after it sends out a message, but it seems not the case. Because it seems that the client sends a notification to server before it closed, so the server still recognizes that session has been disconnected)

Thanks
Hong

**Comment by mrdeep1** [comment]
Hi Hong,

As you are trying to emulate a lossy network, I suggest that you get the client to do a coap_debug_set_packet_loss("%100).  This will prevent the close-down message getting sent by the client. Assuming the client continues to run, then no network unreachable messages will get sent back to the server.

The server will have to be sending a confirmable  request message that will get retried - a CoAP Ping is an example - before it will see that there is a network issue in getting responses back from the client.

**Comment by hongnguyen-tma** [comment]
Hi @mrdeep1 
> I suggest that you get the client to do a coap_debug_set_packet_loss("100%"). 
Can you let me know how to call this function? Because I was thinking of a scenario like this  : client sends request to server > network disconnected > server responses, so if I call this method inside coap-client , the client message won't be able to reach the server. 

I tried by modifying directly in libcoap like this : 
coap_run_once() {
   ....
   timeout = coap_write(ctx, sockets, (unsigned int)(sizeof(sockets) / sizeof(sockets[0])), &num_sockets, before);
   ....
   // My test code starts here
   if (timeout_ms==99) {    // my client set timeout_ms = 99 ms
         if (count == 20){ // 20 is long enough for message in queue get sent
            coap_log(LOG_DEBUG, "Start filtering packet for client....\n");
            coap_debug_set_packet_loss("100%");
         }
    count++;
  }
}
The result is , the client still be able to receive response from server.  Do you know where did I go wrong, or this setting is just for filtering the sent packet, not the received packet? 

>The server will have to be sending a confirmable request message that will get retried - a CoAP Ping is an example - before it will see that there is a network issue in getting responses back from the client.
My server now hasn't had the ability to send Ping message to client. It simply responses to client's request. With this status,  it won't be able to know whether network is disconnected or not, right?

**Comment by mrdeep1** [comment]
coap_debug_set_packet_loss("100%"); is only for the transmission of packets, not the receiving of packets.

As you are wanting to test out the server seeing network issues that triggers a set of packet re-transmits, then there are 2 possible ways I can think of

1) Initiate CoAP ping from server - which you are not set up for

2) Get the client to Observe a resource on the server using a confirmable request.  When the server triggers an observable response, this will need to be confirmable and the client should acknowledge the confirmable response.  With loss set to 100%, then the server will not see the ACK response and should retry sending the confirmable response (see RFC 7641 3.5 Transmission)

**Comment by hongnguyen-tma** [comment]
Hi @mrdeep1 

> Get the client to Observe a resource on the server using a confirmable request. When the server triggers an observable response, this will need to be confirmable and the client should acknowledge the confirmable response

The observable response is the notification itself, isn't it?  If so, it seems that the coap-server can specify either non-confirmable or confirmable for the type of notification message (via function coap_resource_init()). If it uses non-confirmable then notification can't help, right? 
I'm thinking a normal confirmable response (not observable response), it also require client to send ack, which result in retransmit in case of NW disconnected, do you think so? 

Besides, I think that if I set packet loss 100% for coap-server , it won't be able to receive GET request (with Observe Option) from client, so the first response as well as notification won't be sent, right? Do you know how to set it after the server received the request? 

Thanks
Hong

**Comment by hongnguyen-tma** [comment]
Hi @mrdeep1 
Sorry, I might misunderstand the source code.
Current libcoap seeems making the message type of notification message confirmable. 
https://github.com/obgm/libcoap/blob/develop/src/resource.c#L704
However, according to https://tools.ietf.org/html/rfc7641#section-4.5
`A notification can be sent in a confirmable or a non-confirmable
   message.  The message type used is typically application dependent
   and may be determined by the server for each notification
   individually.`

Maybe the current implementation is just one case, it hasn't been completed yet, right?

**Comment by mrdeep1** [comment]
Lines 724-729 update the notification message to non-confirmable if appropriate.  If the request was non-confirmable, the notification message will be non-confirmable.  To get a confirmable notification, thje initial request needs to be confirmable.  This is independent of how coap_resource_init() was set up.

To get the server to re-transmit, either the transmit loss is set to 100% on the server after receiving the conformable request (in the coap_register_handler() registered handler) and before the response is transmitted, or in the client set the transmit loss to 100% in the coap_register_response_handler() handler before the return (so that the ACK response is dropped).

**Comment by mrdeep1** [comment]
Hi Hong,

I would expect the server (in verbose logging) to report the sending of the confirmable response (the underlying TCP/IP stack may drop the packet), but not the client to report seeing it.

What do you mean  by 'without any error'?

**Comment by hongnguyen-tma** [comment]
Hi mrdeep1
I tested with this scenario :  
1. client sends a CON message
2. server receives that message
3. After server creates response (type = ACK) --> make server sleeps for a while
4. Disconnect network in client side before server sends out response. Because client runs in different PC, so I disconnect by plugging off the cable. 
5. Server continue to send response after sleeping time
My expectation :
1. Response cannot reaches client side --> OK
2. Maybe server failed to send the response because network already disconnected --> this was NG.  Actually, the server still shows log message that it sent out xxx bytes, and the message queue is empty when session got closed down.

**Comment by mrdeep1** [comment]
I would expect to see the response message reported by the server as sent in 7 as I mentioned before, the TCP/IP layer would be dropping the packet, not the CoAP layer when you disconnected the network.

However,  the ACK that the client should be sending in response to the confirmable response is not sent  and what is missing is the server not seeing it and therefore triggering the server to re-send the confirmable response max_timeout times.  Correct?

**Comment by hongnguyen-tma** [comment]
Hi mrdeep1
Here is the log in server side : 
Apr 27 08:19:43 DEBG *** [::ffff:10.0.0.162]:4646 <-> [::ffff:103.199.7.226]:34434 (if2) DTLS: session connected
Apr 27 08:19:44 DEBG *  [::ffff:10.0.0.162]:4646 <-> [::ffff:103.199.7.226]:34434 (if2) DTLS: received 76 bytes
v:1 t:CON c:GET i:736a {4264353674693253} [ Uri-Path:.well-known, Uri-Path:test, Uri-Path:v1, Uri-Path:config ]
Apr 27 08:19:44 DEBG call custom handler for resource 0x93758f30

Apr 27 08:19:54 DEBG *  [::ffff:10.0.0.162]:4646 <-> [::ffff:103.199.7.226]:34434 (if2) DTLS: sent 215 bytes ★★★This is the response.  This message was output after I disconnected NW in client side.  The 3 messages go after this message is the session close-notify after timeout.
Apr 27 08:24:54 DEBG *  [::ffff:10.0.0.162]:4646 <-> [::ffff:103.199.7.226]:34434 (if2) DTLS: sent 39 bytes
Apr 27 08:24:54 INFO SSL3 alert write:warning:close notify
Apr 27 08:24:54 DEBG *** [::ffff:10.0.0.162]:4646 <-> [::ffff:103.199.7.226]:34434 (if2) DTLS: session closed

> However, the ACK that the client should be sending in response to the confirmable response is not sent and what is missing is the server not seeing it and therefore triggering the server to re-send the confirmable response max_timeout times. Correct

No, the response from server has type ACK, so I didn't expect the message to be retransmit.  As my original consideration, I would like to know how the server reacts in case network disconnected. Restransmission is a way to know about NW issue and it was straight that it works. So I checked in case of normal message.
